[Diagram] Post 도메인 이벤트 핸들링 다이어그램 작성

codingbaraGo opened the pull request:
<!-- 제목은 `[도메인] PR 제목`으로 작성해주세요. (ex) [User] 소셜 로그인 필터 리팩토링 -->
<!-- PR의 타입은 Label을 통해 나타내주세요. -->

### 💻 작업 내용
<!-- 진행한 작업 내용에 대해 작성해주세요. -->
- Comment 생성/삭제, Post 상세조회 이벤트 핸들링 다이어그램 작성
- 세부 구현 및 기술을 배제한 다이어그램 작성
- 각 이벤트 다이어그램에 핸들링 다이어그램 참조 추가

### ✨ 리뷰 포인트
<!-- (ex) query 가 너무 복잡한 것 같은데 이 위주로 봐주세요. -->
이벤트 다이어그램과 핸들링 다이어그램을 분리한 뒤 !include를 통해 이벤트 다이어그램에 핸들링까지 보이도록 구조를 짰는데, 어떤지 피드백 부탁해요.


### 🎯 관련 이슈
<!-- pr이 merge 되면 이슈가 자동으로 close 되도록 합니다. 만약 자동 close 를 하지 않고 이슈만 링크한다면 closed 키워드를 삭제해주세요.-->
#20

codingbaraGo committed:
docs(diagram): Post 도메인 이벤트 핸들링 다이어그램 작성
- Comment 생성/삭제, Post 상세조회 이벤트 핸들링 다이어그램 작성
- 세부 구현 및 기술을 배제한 다이어그램 작성

codingbaraGo committed:
Merge branch 'develop' of https://github.com/SimpleBoardTeam/simpleBoard-server into document/diagram-handling-post/#20

codingbaraGo committed:
docs(diagram): Post 도메인 이벤트 핸들링 다이어그램 참조 추가
- 각 이벤트 다이어그램에 핸들링 다이어그램 참조 추가

codingbaraGo committed:
chore(diagram): auto-generate SVGs

vwx12345 commented:
게시판(board)가 삭제될 때도 이벤트 처리가 있어야할 것 같습니다.

codingbaraGo commented:
> 게시판(board)가 삭제될 때도 이벤트 처리가 있어야할 것 같습니다.
오! 날카로운 지적 감사합니다.

codingbaraGo committed:
docs(diagram): BoardDeleteHandler 작성

codingbaraGo committed:
Merge branch 'document/diagram-handling-post/#20' of https://github.com/SimpleBoardTeam/simpleBoard-server into document/diagram-handling-post/#20

codingbaraGo committed:
chore(diagram): auto-generate SVGs

codingbaraGo commented:
> 게시판(board)가 삭제될 때도 이벤트 처리가 있어야할 것 같습니다.

고민중인 부분이 BoardDeletedHandler에서 Post와 Comment의 삭제가 모두 이루어져야 할지, Aggregate별로 핸들러를 나누고, 각각의 Aggregate 내부에서 엔티티들을 삭제해야 할지 고민이 되네요

후자를 위해서는 Post -> Comment로 삭제 대상인 postId를 전부 보내줘야하는데 이게 문제에요.
대상 post가 몇만개 이상이면...

vwx12345 commented:
> > 게시판(board)가 삭제될 때도 이벤트 처리가 있어야할 것 같습니다.
> 
> 고민중인 부분이 BoardDeletedHandler에서 Post와 Comment의 삭제가 모두 이루어져야 할지, Aggregate별로 핸들러를 나누고, 각각의 Aggregate 내부에서 엔티티들을 삭제해야 할지 고민이 되네요
> 
> 후자를 위해서는 Post -> Comment로 삭제 대상인 postId를 전부 보내줘야하는데 이게 문제에요. 대상 post가 몇만개 이상이면...

각 Post마다 책임을 지면 되지 않나요? 자신이 삭제될 때마다 post delete 이벤트를 생성하고 그건 comment aggregate 에서 알아서 처리하는 방식으로요

codingbaraGo commented:
> > > 게시판(board)가 삭제될 때도 이벤트 처리가 있어야할 것 같습니다.
> > 
> > 
> > 고민중인 부분이 BoardDeletedHandler에서 Post와 Comment의 삭제가 모두 이루어져야 할지, Aggregate별로 핸들러를 나누고, 각각의 Aggregate 내부에서 엔티티들을 삭제해야 할지 고민이 되네요
> > 후자를 위해서는 Post -> Comment로 삭제 대상인 postId를 전부 보내줘야하는데 이게 문제에요. 대상 post가 몇만개 이상이면...
> 
> 각 Post마다 책임을 지면 되지 않나요? 자신이 삭제될 때마다 post delete 이벤트를 생성하고 그건 comment aggregate 에서 알아서 처리하는 방식으로요

삭제 대상 게시판에 게시글이 수십만 ~ 수백만개가 존재한다면 그 개수대로 이벤트가 발생하게 되고, Comment에서는 이벤트 개수만큼의 삭제 쿼리를 실행시키게 됩니다.
한번에 그런 개수의 이벤트가 발생한다면 이벤트 퍼블리셔 혹은 DB 둘중 하나에는 성능문제가 발생할 수 있을것같아요.

vwx12345 commented:
> > > > 게시판(board)가 삭제될 때도 이벤트 처리가 있어야할 것 같습니다.
> > > 
> > > 
> > > 고민중인 부분이 BoardDeletedHandler에서 Post와 Comment의 삭제가 모두 이루어져야 할지, Aggregate별로 핸들러를 나누고, 각각의 Aggregate 내부에서 엔티티들을 삭제해야 할지 고민이 되네요
> > > 후자를 위해서는 Post -> Comment로 삭제 대상인 postId를 전부 보내줘야하는데 이게 문제에요. 대상 post가 몇만개 이상이면...
> > 
> > 
> > 각 Post마다 책임을 지면 되지 않나요? 자신이 삭제될 때마다 post delete 이벤트를 생성하고 그건 comment aggregate 에서 알아서 처리하는 방식으로요
> 
> 삭제 대상 게시판에 게시글이 수십만 ~ 수백만개가 존재한다면 그 개수대로 이벤트가 발생하게 되고, Comment에서는 이벤트 개수만큼의 삭제 쿼리를 실행시키게 됩니다. 한번에 그런 개수의 이벤트가 발생한다면 이벤트 퍼블리셔 혹은 DB 둘중 하나에는 성능문제가 발생할 수 있을것같아요.

무슨 말인지 이해했습니다. 하지만 comment는 board랑 직접 연관이 없는데 한번에 처리하는 건 현재 도메인 구조로는 불가능할 것 같습니다.
그래서 둘 중 하나를 선택해야할 것 같네요.
1. 도메인을 변경해서 board delete Event를 comment에서 처리하게 한다.
2. post에서 지워진 post Id들을 넘기는 posts delete Event로 처리한다.

codingbaraGo commented:
> > > > > 게시판(board)가 삭제될 때도 이벤트 처리가 있어야할 것 같습니다.
> > > > 
> > > > 
> > > > 고민중인 부분이 BoardDeletedHandler에서 Post와 Comment의 삭제가 모두 이루어져야 할지, Aggregate별로 핸들러를 나누고, 각각의 Aggregate 내부에서 엔티티들을 삭제해야 할지 고민이 되네요
> > > > 후자를 위해서는 Post -> Comment로 삭제 대상인 postId를 전부 보내줘야하는데 이게 문제에요. 대상 post가 몇만개 이상이면...
> > > 
> > > 
> > > 각 Post마다 책임을 지면 되지 않나요? 자신이 삭제될 때마다 post delete 이벤트를 생성하고 그건 comment aggregate 에서 알아서 처리하는 방식으로요
> > 
> > 
> > 삭제 대상 게시판에 게시글이 수십만 ~ 수백만개가 존재한다면 그 개수대로 이벤트가 발생하게 되고, Comment에서는 이벤트 개수만큼의 삭제 쿼리를 실행시키게 됩니다. 한번에 그런 개수의 이벤트가 발생한다면 이벤트 퍼블리셔 혹은 DB 둘중 하나에는 성능문제가 발생할 수 있을것같아요.
> 
> 무슨 말인지 이해했습니다. 하지만 comment는 board랑 직접 연관이 없는데 한번에 처리하는 건 현재 도메인 구조로는 불가능할 것 같습니다. 그래서 둘 중 하나를 선택해야할 것 같네요.
> 
> 1. 도메인을 변경해서 board delete Event를 comment에서 처리하게 한다.
> 2. post에서 지워진 post Id들을 넘기는 posts delete Event로 처리한다.

Post와 Comment는 같은 B.C이니 도메인 서비스 클래스가 양쪽 레포지토리를 가져와 처리하는 방법은 어떻게 생각하세요?
쿼리 하나에 join을 통해 Post와 Comment 모두 삭제하는 것 까지는 힘들더라도 이벤트를 통한 전달 없이 양쪽 레포지토리에 삭제 명령을 내리면 오버헤드가 줄어들 듯 한데 어떨까요?

vwx12345 commented:
> > > > > > 게시판(board)가 삭제될 때도 이벤트 처리가 있어야할 것 같습니다.
> > > > > 
> > > > > 
> > > > > 고민중인 부분이 BoardDeletedHandler에서 Post와 Comment의 삭제가 모두 이루어져야 할지, Aggregate별로 핸들러를 나누고, 각각의 Aggregate 내부에서 엔티티들을 삭제해야 할지 고민이 되네요
> > > > > 후자를 위해서는 Post -> Comment로 삭제 대상인 postId를 전부 보내줘야하는데 이게 문제에요. 대상 post가 몇만개 이상이면...
> > > > 
> > > > 
> > > > 각 Post마다 책임을 지면 되지 않나요? 자신이 삭제될 때마다 post delete 이벤트를 생성하고 그건 comment aggregate 에서 알아서 처리하는 방식으로요
> > > 
> > > 
> > > 삭제 대상 게시판에 게시글이 수십만 ~ 수백만개가 존재한다면 그 개수대로 이벤트가 발생하게 되고, Comment에서는 이벤트 개수만큼의 삭제 쿼리를 실행시키게 됩니다. 한번에 그런 개수의 이벤트가 발생한다면 이벤트 퍼블리셔 혹은 DB 둘중 하나에는 성능문제가 발생할 수 있을것같아요.
> > 
> > 
> > 무슨 말인지 이해했습니다. 하지만 comment는 board랑 직접 연관이 없는데 한번에 처리하는 건 현재 도메인 구조로는 불가능할 것 같습니다. 그래서 둘 중 하나를 선택해야할 것 같네요.
> > 
> > 1. 도메인을 변경해서 board delete Event를 comment에서 처리하게 한다.
> > 2. post에서 지워진 post Id들을 넘기는 posts delete Event로 처리한다.
> 
> Post와 Comment는 같은 B.C이니 도메인 서비스 클래스가 양쪽 레포지토리를 가져와 처리하는 방법은 어떻게 생각하세요? 쿼리 하나에 join을 통해 Post와 Comment 모두 삭제하는 것 까지는 힘들더라도 이벤트를 통한 전달 없이 양쪽 레포지토리에 삭제 명령을 내리면 오버헤드가 줄어들 듯 한데 어떨까요?

그럼
1. Post를 통해서 지워야할 comment를 식별하는 방식
2. Board를 통해서 지워야할 comment를 식별하는 방식
둘 중 어떤 방식으로 구현하실건가요? 아니면 또 다른 방식으로 구현 계획이 있으신가요?

codingbaraGo commented:
> > > > > > > 게시판(board)가 삭제될 때도 이벤트 처리가 있어야할 것 같습니다.
> > > > > > 
> > > > > > 
> > > > > > 고민중인 부분이 BoardDeletedHandler에서 Post와 Comment의 삭제가 모두 이루어져야 할지, Aggregate별로 핸들러를 나누고, 각각의 Aggregate 내부에서 엔티티들을 삭제해야 할지 고민이 되네요
> > > > > > 후자를 위해서는 Post -> Comment로 삭제 대상인 postId를 전부 보내줘야하는데 이게 문제에요. 대상 post가 몇만개 이상이면...
> > > > > 
> > > > > 
> > > > > 각 Post마다 책임을 지면 되지 않나요? 자신이 삭제될 때마다 post delete 이벤트를 생성하고 그건 comment aggregate 에서 알아서 처리하는 방식으로요
> > > > 
> > > > 
> > > > 삭제 대상 게시판에 게시글이 수십만 ~ 수백만개가 존재한다면 그 개수대로 이벤트가 발생하게 되고, Comment에서는 이벤트 개수만큼의 삭제 쿼리를 실행시키게 됩니다. 한번에 그런 개수의 이벤트가 발생한다면 이벤트 퍼블리셔 혹은 DB 둘중 하나에는 성능문제가 발생할 수 있을것같아요.
> > > 
> > > 
> > > 무슨 말인지 이해했습니다. 하지만 comment는 board랑 직접 연관이 없는데 한번에 처리하는 건 현재 도메인 구조로는 불가능할 것 같습니다. 그래서 둘 중 하나를 선택해야할 것 같네요.
> > > 
> > > 1. 도메인을 변경해서 board delete Event를 comment에서 처리하게 한다.
> > > 2. post에서 지워진 post Id들을 넘기는 posts delete Event로 처리한다.
> > 
> > 
> > Post와 Comment는 같은 B.C이니 도메인 서비스 클래스가 양쪽 레포지토리를 가져와 처리하는 방법은 어떻게 생각하세요? 쿼리 하나에 join을 통해 Post와 Comment 모두 삭제하는 것 까지는 힘들더라도 이벤트를 통한 전달 없이 양쪽 레포지토리에 삭제 명령을 내리면 오버헤드가 줄어들 듯 한데 어떨까요?
> 
> 그럼
> 
> 1. Post를 통해서 지워야할 comment를 식별하는 방식
> 2. Board를 통해서 지워야할 comment를 식별하는 방식
>    둘 중 어떤 방식으로 구현하실건가요? 아니면 또 다른 방식으로 구현 계획이 있으신가요?

둘중 어느 aggregate도 아닌, Board B.C아래에 두는게 맞지 않을까 합니다.

Handler가 직접 처리하는 방식과,
Board B.C 아래에 별도의 도메인 서비스를 추가해서 Handler 가 domain service를 호출하는 방식 중 고민중입니다.

codingbaraGo committed:
docs(diagram): Board deleted handling 다이어그램 수정
- Comment aggregate에 board 정보를 넣어 대량 수정시 Post 목록 조회 없이, boardId를 조건으로 삭제 가능하도록 변경

codingbaraGo committed:
[Merge]: develop <- document/diagram-handling-post/#20

codingbaraGo committed:
chore(diagram): auto-generate SVGs
